### ISSUE TITLE
Leaking Event handlers when using multiple elm apps with routing on page

### ISSUE THREAD

**Issue by ChristophP** [elm-lang/core#886]
## TL;DR

When having multiple elm apps using `Navigation.program` in an SPA you can't stop an app from listening to url changes even when they aren't currently displayed. This can lead to weird behavior forces workarounds inside the apps and could lead to memory leaks and performance issues.

All workarounds target "ignoring" the incoming events instead of actually unbinding their handles. Having an `app.kill()` function would be the cleanest solution for the issue at hand, but also be beneficial in other unrelated situations.
- no workaround code needed inside the apps.
- more control over the lifetime of the app and avoid possible memory leaks. This could be a biggie. Leaking memory and/or having handlers running in the background is something that might lead to hard-to-debug behavior in single-page-apps.
- `.embed()` would get a complementary destroy function and not feel so lonely anymore :-)

## The problem

We have an app shell architecture, meaning a host app that loads several other apps on our page. This lets us maintain and deploy each app separately. The app shell routes to `/app1`  and App 1 will route to everything after `/app1/...`. Some of the loaded apps are using `Navigation.program`. When the path changes from `/app1` to `/app2`, App 1 still listens to popstate events even though it should be idle.

## Possible workarounds

There are a couple of ways to circumvent this from being a problem.
1. Removing all subscriptions. => Won't work. There is no way to tell the `Navigation.program` to stop listening to this event since the subscription is untouchable and always there https://github.com/elm-lang/navigation/blob/master/src/Navigation.elm#L73.
2. "Turning off" the App1 when routing away (with something like this `update = (\model msg -> if model.isMounted then update model msg else (model, Cmd.none))`) => would work but bloats the model and update function a little bit.
3. Making sure each app only parses routes that are within it's namespace (only parse stuff starting with `/app1`)

All these methods will just elimate the reaction to the event, rather than removing the actual event listener.

## Proposed solutions

I think 2 solutions would solve this problem, of which the second is preferable.
1. Having some way of telling the Navigation.program to stop listening to popstate(would require ports to notify the program to stop/start listening)
2. Actually having an `app.kill()` function which destroys the elm app and unbinds all handlers and maybe cleans up the DOM.

## Also ...
Having `app.kill()` would also be benficial in a different scenario (correct me if I am wrong):
Say I am using `react-elm-components` and unmount and remount my component a couple of times by navigating away and back to it. Every time an elm app is created and possibly attaches handlers to the page. Simply calling `app.kill` in `componentWillUnmount` would solve that issue.

**Comment by ChristophP** [comment]
I actually found that that leaking stuff is not necessarily specific to multiple elm apps on the page, but more to reinitializing elm modules(calling `Elm.<Module>.embed/fullscreen()`). This can happen for example when using `react-elm-components` and mounting the component multiple times via navigating through the page if subscriptions are used without reloading the page.

I created an example here: https://www.google.com/url?q=https%3A%2F%2Fellie-app.com%2F3LHDqnYqD4za1%2F10&sa=D&sntz=1&usg=AFQjCNFJOQ1jz9_IdoqIEyce6Z7_YIv1hA

**Comment by joefiorini** [comment]
@ChristophP Almost a year later, but I just came up with a workaround for this issue; using web components, you can fire a custom event when the elm app is detached from the DOM. You could then create a custom program function that automatically disconnects subscriptions when the app is unmounted. I have not tested this extensively yet, but I can verify it fixes the ellie example you posted above. Here's the code:

https://gist.github.com/joefiorini/4372e2362e658d3092419e708416f92c

Sorry it's not an ellie, but it was easier to write in an editor with multiple files due to having multiple `Msg` and `Model` types and the custom element.

**Comment by Pilatch** [comment]
Adding workaround for the memory leak where smaller Elm apps embedded in other things - https://github.com/Pilatch/angularjs-ng-elm#technical-stuff

**Comment by ChristophP** [comment]
@joefiorini I haven't thought of using web components callbacks yet but nice idea. I have been using something similar with ports and callbacks to clean up lingering subscriptions. But it would we cool if there was a function for elm that clobbers everything down on command. However now with 0.19 out, ther only seems to be a string use case for this with `Browser.element`. The other ones will pretty much always live as long as the page does.

**Comment by AbGuthrie** [comment]
+1 Elm applications are a great candidate for stable drop-in components, and having no way to cleanup lingering listeners is very detrimental.

**Comment by FranzSkuffka** [comment]
Is this still a thing in 0.19?

**Comment by ChristophP** [comment]
@FranzSkuffka Yup

**Comment by winduptoy** [comment]
Here's another use case:

I'm using Elm in a browser extension content script, so I inject a little Elm app on certain pages. This is happening on someone else's single-page-app, so when the user navigates away from the current view, I need to clean up my Elm stuff.

**Comment by LimmaPaulus** [comment]
Up! I have a very similar situation as @ChristophP have.

**Comment by Janiczek** [comment]
I second the need/want for a way to completely stop, unsubscribe, ..., cleanup the Elm app on demand. My usecase is similar - SPA composed of various sub-apps (Elm, React, etc.).

**Comment by toastal** [comment]
I would suggest this being called `terminate()` instead of `kill()`. This mimics the Worker API and sounds less violent.

**Comment by ChristophP** [comment]
I just came up with another reason why a stop, unsubscribe, destroy mechanism for an elm app/element would be good:
**the debugger**
The debugger records a history of all events that happened. In a scenario with multiple pages and `Browser.element`s running on different pages each one will have a debugger instance.Now when on pageA a debugger will be spawned, when navigating to page B a new debugger will be spawned. When navigating back to page A, yet another debugger for page A we be created and the old one is still in memory. Depending on the frequency is size of messages this leaks memory over time in development.

**Comment by ChristophP** [comment]
So all in all what a `app.kill/app.destroy` should do to accomodate the scenarios for mulitple `Browser.element`s on the page:
1. remove all subscriptions
2. kill the debugger/clear it's history
3. unbind all ports code
4. unmount things from the DOM
5. clean up any other things that may linger in memory
6. (not necessary IMO if the other things work) optionally allow for a callback to react to the destruction

**Comment by janwirth** [comment]
Noting here a reference implementation by @supermario 
https://gist.github.com/supermario/4c2615806c6c561a16edf5dd7208a759#js-modification-solution

**Comment by janwirth** [comment]
I just opened #1125 and I used the name `shutdown` (the name of the commit is still incorrect). I was looking for nicer words than "destroy", "terminate", "kill".

I considered:
- "close", but elm also can be used to build headless applications. You would not "close" a service, but you can shut down pretty anything.
- "end" similar to "close" you would end a game, but not an app or a service
- "stop" was also considered, but it suggests that we suspend something which could be resumed.
- "disconnect" is similar to "stop". it hints at something that can be reconnected
- something that I "shut down" can be turned on again - but it loses its in-memory state.

**Comment by ChristophP** [comment]
Very interesting. And thanks for sharing the gist.
